app/src/main/java/org/thoughtcrime/securesms/webrtc/CallMessageProcessor.kt

```diff
@@ -27,28 +29,58 @@ import org.session.libsignal.protos.SignalServiceProtos.CallMessage.Type.PROVISI
 
     companion object {
         private const val VERY_EXPIRED_TIME = 15 * 60 * 1000L
 
-        fun safeStartService(context: Context, intent: Intent) {
-            // If the foreground service crashes then it's possible for one of these intents to
-            // be started in the background (in which case 'startService' will throw a
-            // 'BackgroundServiceStartNotAllowedException' exception) so catch that case and try
-            // to re-start the service in the foreground
+        fun safeStartForegroundService(context: Context, intent: Intent) {
+            // Attempt to start the call service..
             try { context.startService(intent) }
             catch(e: Exception) {
+                // While fine if the app is in the foreground, you cannot start a foreground service easily
+                // in Android 12 and above if the device is locked (i.e., if you get a call when the device is
+                // locked & attempt start the foreground service) it will throw aa BackgroundServiceStartNotAllowedException such as:
+                //      Unable to start CallMessage intent: startForegroundService() not allowed due to mAllowStartForeground false:
+                //      service network.loki.messenger/org.thoughtcrime.securesms.service.WebRtcCallService
+                //
+                // ..as such, we'll wake the device up before attempting to start the foreground service.
+                wakeUpDeviceAndDismissKeyguard(context)
                 try { ContextCompat.startForegroundService(context, intent) }
                 catch (e2: Exception) {
                     Log.e("Loki", "Unable to start CallMessage intent: ${e2.message}")
                 }
             }
         }
+
+        // Wake the device up if it's asleep / locked (used when we receive an incoming call)
+        private fun wakeUpDeviceAndDismissKeyguard(context: Context) {
+            val powerManager = context.getSystemService(Context.POWER_SERVICE) as PowerManager
+            val keyguardManager = context.getSystemService(Context.KEYGUARD_SERVICE) as KeyguardManager
+
+            // Check if the screen is awake & if the phone is locked
+            val isPhoneLocked = keyguardManager.isKeyguardLocked
+            val isScreenAwake = powerManager.isInteractive
+
+            if (!isScreenAwake) {
+                val wakeLock = powerManager.newWakeLock(
+                    PowerManager.FULL_WAKE_LOCK or
+                            PowerManager.ACQUIRE_CAUSES_WAKEUP or
+                            PowerManager.ON_AFTER_RELEASE,
+                    "${NonTranslatableStringConstants.APP_NAME}:WakeLock"
+                )
+
+                // We only need the wake lock briefly
+                wakeLock.acquire(3000)
+            }
+
+            if (isPhoneLocked) {
+                val keyguardLock = keyguardManager.newKeyguardLock("{${NonTranslatableStringConstants.APP_NAME}:KeyguardLock")
```

Comment:
Should this constant be extracted and reused? `"{${NonTranslatableStringConstants.APP_NAME}:KeyguardLock"`

Reply:
Done (moved to ApplicationContext and re-used whenever required)